lib/rugged_adapter/git_layer_rugged.rb

```diff
@@ -165,22 +165,25 @@ def exist?
         ::File.exists?(@repo.path)
       end
 
-      def grep(query, options={})
-        ref = options[:ref] ? options[:ref] : "HEAD"
-        tree = @repo.lookup(sha_from_ref(ref)).tree
-        tree = @repo.lookup(tree[options[:path]][:oid]) if options[:path]
-        enc = options.fetch(:encoding, 'utf-8')
+      def grep(search_terms, options={})
+        ref   = options[:ref] ? options[:ref] : "HEAD"
+        tree  = @repo.lookup(sha_from_ref(ref)).tree
+        tree  = @repo.lookup(tree[options[:path]][:oid]) if options[:path]
+        enc   = options.fetch(:encoding, 'utf-8')
+        query = /^(.*(?:#{search_terms.join('|')}).*)$/i
         results = []
         tree.walk_blobs(:postorder) do |root, entry|
-          blob = @repo.lookup(entry[:oid])
+          blob  = @repo.lookup(entry[:oid])
           count = 0
-          next if blob.binary?
-          blob.content.force_encoding(enc).each_line do |line|
-            next unless line.match(/#{query}/i)
-            count += 1
+          count += 1 if entry[:name] =~ query
```

Comment:
Before we were searching for filename matches in a second loop through the repo in `gollum-lib`, which is just silly.